README.md

```diff
@@ -1644,6 +1645,60 @@ variables](https://www.terraform.io/docs/configuration/environment-variables.htm
 prefix `TF_VAR_`, so one way to share a variable named `foo` between Terraform and Terragrunt is to set its value
 as the environment variable `TF_VAR_foo` and to read that value in using this `get_env()` built-in function.
 
+#### get_output
+
+`get_output(TARGET_TERRAGRUNT_CONFIG, [NAME])` returns the value of the output of the terraform module configured with
+the terragrunt config specified at `TARGET_TERRAGRUNT_CONFIG`. If provided a `NAME`, this function will return just the
+output indexed by `NAME`. Otherwise, this will return the all outputs as a map. The return type is preserved. For
+example, if the output is of type `number`, the returned value of `get_output` will also be of type `number`.
+
+Example:
+
+Suppose you had the following terragrunt live folder structure:
+
+```
+/terraform-code
+├── app1
+│   └── terragrunt.hcl
+├── tests
+│   ├── app2
+│   |   └── terragrunt.hcl
+│   └── app3
+│       └── terragrunt.hcl
+```
+
+Suppose `app1` had the following outputs:
+
+```
+output "instance_id" {}
+output "instance_name" {}
+```
+
+In `app3`, you can get each of these outputs using `get_output` in the following ways:
+
+```hcl
+inputs = {
+  instance_id_directly = get_output("${get_terragrunt_dir()}/../../app1", "instance_id")
+  instance_name_directly = get_output("${get_terragrunt_dir()}/../../app1", "instance_name")
+  instance_id_indirectly = get_output("${get_terragrunt_dir()}/../../app1")["instance_id"]
+}
+```
+
+Note that the terragrunt run configuration is inherited from the current session. What this means is that settings set
+with environment variables (such as `TERRAGRUNT_IAM_ROLE`) are automatically passed down to the execution context of
+the target terragrunt config when reading the output data.
+
+**Important**: Because `get_output` depends on the target config having been applied already, you will need to make sure
+that the specified `TARGET_TERRAGRUNT_CONFIG` is listed as a dependency in the `dependencies` block. Otherwise, this
+will fail when you run an `apply-all` since the output will be interpolated before the target config has applied.
+Additionally, during an `apply-all`, the terragrunt configuration has to be partially interpolated in order to build up
+the dependency tree. This means that you will have issues using `apply-all` if `get_output` is used in the following
+blocks, as it will be interpolated prior to any modules being applied:
+
+- `locals`
```

Comment:
Hm, if you need to fetch 10 different outputs from a single dependency, I suspect 10 sequential `terragrunt output` calls would be slow. Do we cache the outputs in memory in between? If not, it would be nice to allow users to cache all the outputs in a `locals` variable once and then look up values within it...

Reply:
Yea that is a good point. Under most circumstances, you can use `get_output` in `locals`. There are some caveats you have to be aware of: https://github.com/gruntwork-io/terragrunt/pull/828#discussion_r312662399

Reply:
My first impression was that we should maintain an internal, in-memory cache of (module absolute path, all outputs for that module) pairs. Any time you call `get_output`, it first checks the cache for that value: if it's present, it returns it immediately; if it's absent, it calls `terraform output` and caches the result. This ensures maximal performance without any extra effort for the user...

Reply:
Depending on the implementation, I worry about what happens when we are going through multiple passes of the config, as it does in the `xxx-all` command. Here is where a global memoized cache could be problematic:

- In the first pass through, it would run `get_output` on the *previous* state. This result gets cached.
- The dependency gets applied as part of `apply-all`, in the dependency order.
- We parse the config again to apply the dependent module. As part of that, we call `get_output`. We look up in the cache, and return the previous value, which is NOT the newly applied state.

Note that this only happens when you depend on the awkward chicken-and-egg scenario of using `get_output` in `locals`, and in a naive implementation using a global cache that is stored for the duration of a single `terragrunt` call.

We could have a different implementation where the cache is scoped to a single config parsing, which trades off more reuse for better predictability. I am leaning towards implementing it this way, as cache bugs are always super hard to debug. This of course requires more refactoring, since now I need to store the cache somewhere in a struct that is available throughout the parsing stage.

Reply:
In the new implementation, you will call `terragrunt output` at most once per dependency per config. There is still room for memoization to cache across `apply-all` calls, but I think this is world's better than `get_output` which was once per function call.

And we don't need caching (which is always a pain to maintain due to annoying cache busting bugs) :)